### ISSUE TITLE
lg2 merge exit code incorrect

### ISSUE THREAD

**Issue by RuthlessJailer** [holzschu/a-shell#667]
When testing a simple sh script I noticed a very frustrating error. The same script behaves differently when there is a merge conflict. Namely, when you see this output:
```
Received 0/0 objects in 0bytes
Error: repository is in unexpected state:
repository is in state 1
It looks like a merge is in progress. Either 
resolve the conflicts (see `lg2 status`), `lg
2 add` each changed file and commit the resul
t, or run `lg2 reset --hard HEAD` to stop the
 merge.
```
In this case, status code 0 is apparently returned, since my if statement prints "succ". However, if I run the same file in a folder without a repo, it prints "fail".

Here is the aforementioned script:
```sh
#! /bin/sh
echo hey 
if ! ( git pull ); then
        echo fail
else
        echo succ
fi
```

Hopefully there can be a small patch, but if not or if I'm doing something wrong, let me know and I'll understand!

**Comment by RuthlessJailer** [comment]
Notice: I used `pull` as an example but I tend to use `fetch` and `merge` which yields the same result.

**Comment by holzschu** [comment]
That's an interesting issue. I think I tracked it to this line, which always returns `0` instead of `err`: 
https://github.com/holzschu/libgit2/blob/7241be4f29bd44fd56be5f75e5a138e9303ca39f/examples/merge.c#L385

`fetch` and `pull` are different, since they return `-1` in case of an error. I'll have to investigate more.

**Comment by RuthlessJailer** [comment]
Hey, thanks for looking into this. It's encouraging to see how dedicated you are tot this project!

**Comment by holzschu** [comment]
Okay, this may be a stupid question, but... how can I intentionally put a repository in a state where "git pull" fails? 
I've done it thousands of times not on purpose, but now that I need to do it, I fail.

**Comment by RuthlessJailer** [comment]
So I've been messing with this on and off the past few days and I have sort of found the issue. My setup is I'm trying to sync Obsidian notes between PC and iPhone via an ssh repo on raspberrrpi.local. When this works, it's pretty cool, however, it seems to not like it when there are merge conflicts (and those are annoying to resolve without mergetool).

So my repo was kind of messed up, but here is a fresh diagnosis.
1) Create a new GitHub repo (`/RuthlessJailer/test` will be used in this instruction set, however you can set up SSH if you wish)
2) Clone the repo
3) cd into the cloned folder
4) Setup the cred:
```bash
git config user.name a-shell
git config user.email iOS
git config user.password <your_github_token>
# or if using ssh, set password to " ", and:
# git config user.IdentityFile /path/to/sk_id
```
5) Create some dummy data;
```bash
touch file
date > stat
```
6) commit & push
```bash
git add .
git commit -m "push"
git push
```
7) Intentionally create a merge conflict by editing a file (on GitHub, or a clone of your repo somewhere else):
```bash
echo EVIL MERGE CONFLICT > stat
# commit & push
```
8) Edit the same file on iOS **BEFORE** you pull
```bash
date > stat
# only commit
```
9) Now `git pull` twice:
```
[test]$ git pull
Fetching origin for repo 0x1075855b0
Connecting to github.com...
Received 0/0 objects in 0bytes
conflict: a:stat o:stat t:stat
[test]$ git pull
Fetching origin for repo 0x1077210e0
Connecting to github.com...
Received 0/0 objects in 0bytes
Error: repository is in unexpected state:
repository is in state 1
It looks like a merge is in progress.
Either resolve the conflicts (see `lg2 status`), `lg2 add` each changed file and commit the result, or run `lg2 reset --hard HEAD` to stop the merge.
[test]$
```
10) Now, if you do steps 7-9 again, but instead my original script in this issue, both types it prints "succ".

---
Now, sort of unrelated to this issue, but there is a state in which lg2 consistently prints `segmentation fault` upon pull (each pull immediately after `lg2 reset --hard`). I do not yet know how to reproduce this, but I will open a future issue once I have figured it out.
[IMG_5099](https://github.com/holzschu/a-shell/assets/57777932/e6c93168-a4a9-4374-803f-110df1f04650)

---
Once again, appreciate your commitment to this project!

**Comment by holzschu** [comment]
Hi, 
Thanks for your help. Build 344 of a-Shell mini should have fix for lg2 pull and merge. I'm not sure about fetch.

It is on TestFlight, so you can experiment with it before I make the full release:
https://testflight.apple.com/join/REdHww5C

**Comment by tabedzki** [comment]
What is the status on this? As of Jan 2024, I still experience the segfaults and errors described above.

**Comment by holzschu** [comment]
The main issue was that lg2 did not report the correct error code. That one has been fixed and released.
For the segfault issue, I need a way to reproduce it in order to fix it.

**Comment by tabedzki** [comment]
I have the segfault when I do a lg2 pull. That fails and leads to the exit status 1 and locks it. What's the current version of lg2? Mine reports 1.1.0 for the up to date App Store (Jan 25, 2024).

**Comment by holzschu** [comment]
The key question is: how can I reproduce this issue?
In order to indentify the cause of the problem (and fix it), I need to reproduce it and analyze it.

**Comment by tabedzki** [comment]
Agreed. I'd like to confirm that I'm using the latest lg2 version available before I start trying to replicate this issue. Is 1.1.0 the latest version?

**Comment by holzschu** [comment]
It's the latest in a-Shell.

**Comment by sol-sk** [comment]
Is this issue still unresolved? I'm getting the same error message, also trying to sync Obsidian iOS with desktop.

**Comment by holzschu** [comment]
My question is going to be the same as one year ago: I need a way to reproduce the issue in order to (attempt to) fix it.

Also "this issue" is not enough information, as there are two issues in this thread: lg2 pull and merge in some situations, and lg2 doing a sefmentation fault in other situations.

**Comment by justinmklam** [comment]
FWIW I was running into the same segmentation fault issue when my local branch had merge conflicts with remote. The error:

```
$ lg2 pull
Fetching origin for repo 0x1026147a0
Host: github.com
Public key hashes:
MD5: ***
SHA-1: ***
SHA-256: ***
There are 1 known hosts...
Host github.com is in known_hosts!
SSH authentication: Using private key: ~/Docume
nts/.ssh/id_ed25519
Received 0/0 objects in 0bytes
Fetching origin for repo 0x1026147a0

Host: github.com
Public key hashes:
MD5: ***
SHA-1: ***
SHA-256: ***

There are 1 known hosts...
Host github.com is in known_hosts!
SSH authentication: Using private key: ~/Documents/.ssh/id_ed25519
Received 0/0 objects in 0bytes
segmentation fault
```

The other error I was getting was:

```
[personal-blog]$ lg2 checkout draft/now
ERROR: 
repository is in state 1
It looks like a merge is in progress. Either re
solve the conflicts (see `lg2 status`), `lg2 ad
d` each changed file and commit the result, or 
run `lg2 reset --hard HEAD` to stop the merge.
```

I then ran what it suggested, which looked promising but still did not pull successfully from main: 

```
[personal-blog]$ lg2 reset --hard HEAD
[personal-blog]$ lg2 status
# On branch main
```

Instead, I ran this:

```
lg2 reset --hard origin/main
```
Which successfully reset my local repo to remote and fixed the weird state my repo got in.
